core/main/src/broker/endpoint_broker.rs

```diff
@@ -0,0 +1,598 @@
+// Copyright 2023 Comcast Cable Communications Management, LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use ripple_sdk::{
+    api::{
+        firebolt::fb_capabilities::JSON_RPC_STANDARD_ERROR_INVALID_PARAMS,
+        gateway::rpc_gateway_api::{
+            ApiMessage, ApiProtocol, CallContext, JsonRpcApiResponse, RpcRequest,
+        },
+        session::AccountSession,
+    },
+    extn::extn_client_message::ExtnMessage,
+    framework::RippleResponse,
+    log::error,
+    tokio::{
+        self,
+        sync::mpsc::{Receiver, Sender},
+    },
+    utils::error::RippleError,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Value};
+use std::{
+    collections::HashMap,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc, RwLock,
+    },
+};
+
+use crate::{
+    firebolt::firebolt_gateway::JsonRpcError,
+    state::platform_state::PlatformState,
+    utils::router_utils::{return_api_message_for_transport, return_extn_response},
+};
+
+use super::{
+    http_broker::HttpBroker,
+    rules_engine::{jq_compile, Rule, RuleEndpoint, RuleEndpointProtocol, RuleEngine},
+    thunder_broker::ThunderBroker,
+    websocket_broker::WebsocketBroker,
+};
+
+#[derive(Clone, Debug)]
+pub struct BrokerSender {
+    pub sender: Sender<BrokerRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct BrokerRequest {
+    pub rpc: RpcRequest,
+    pub rule: Rule,
+}
+
+/// BrokerCallback will be used by the communication broker to send the firebolt response
+/// back to the gateway for client consumption
+#[derive(Clone, Debug)]
+pub struct BrokerCallback {
+    pub sender: Sender<BrokerOutput>,
+}
+
+static ATOMIC_ID: AtomicU64 = AtomicU64::new(0);
+
+impl BrokerCallback {
+    /// Default method used for sending errors via the BrokerCallback
```

Comment:
Is this is the "default" method used to send errors does that mean there's a way to override it with another method?

Reply:
This is an utility method would be useful as a last resort for
1. Invalid Rule
2. Invalid Endpoint
3. Rule compilation errors
4. Rule Filter errors
5. Timeout cases(future)